// @rb-ui/frontend.kit-npm/src/frontend-kit/atoms/link/link.tsx
import * as React from 'react';
import classNames from 'classnames';

class LinkProps {
  level: 'simple' | 'primary' | 'inline';
  href: string;
  label: string;
  disabled?: boolean;
  target?: '_blank' | '_self';
}

/**
 * @name    a-link
 * @type    atom
 * @author Experience One AG
 * @copyright Robert Bosch GmbH
 *
 * @param   {string} level        Type of Link (Primary or Secondary)
 * @param   {string} href         HREF-Attribute
 * @param   {string} label        Label to Display
 * @param   {boolean} showArrow   Show or Hide Arrow-Icon after Label
 * @param   {boolean} disabled    Whether or not the Link is disabled
 * @param   {string} target       Open Links in same or new window (_self or _blank)
 * @description
 * representation of links
 */
const Link: React.FunctionComponent<LinkProps> = ({
  level = 'simple',
  href,
  label,
  disabled,
  target = '_self',
}: LinkProps) => {
  // construction of class names
  const linkClass = classNames('a-link', {
    'a-link--primary': level === 'primary',
    'a-link--simple': level === 'simple',
    '-disabled': disabled,
  });

  const isInlineLink = (level === 'inline');

  if (isInlineLink) {
    return (
      <a href={href} target={target}>
        {label}
      </a>
    );
  }

  return (
    <div className={linkClass}>
      <a href={href} target={target}>
        {label}
      </a>
    </div>
  );
  
};

export { Link, LinkProps };
