// @rb-ui/frontend.kit-npm/src/frontend-kit/atoms/searchfield/searchfield.tsx
import * as React from 'react';
import classNames from 'classnames';

class SearchfieldProps {
  disabled?: boolean;
  id: string;
  placeholder?: string;
  dynamicWidth?: boolean;
}

/**
 * @name        a-searchfield
 * @type        atom
 * @author      Experience One AG
 * @copyright   Robert Bosch GmbH
 *
 * @param       disabled      Wether or not the search field is disabled
 * @param       id              Unique ID for the search field
 * @param       placeholder     Placeholder of search field
 * @param       dynamicWidth    Wether or not the search field has a dynamic width
 * @description
 * representation of search field
 */

const Searchfield: React.FunctionComponent<SearchfieldProps> = ({
  disabled,
  id,
  placeholder,
  dynamicWidth,
}: SearchfieldProps) => {
  const inputClass = classNames('a-searchfield', {
    [`a-searchfield--dynamic-width`]: dynamicWidth,
  });

  return (
    <div className={inputClass}>
      <input
        type="search"
        id={id}
        disabled={disabled}
        placeholder={placeholder}
      />
      <label htmlFor={id} />
    </div>
  );
};

export { Searchfield, SearchfieldProps };
