// @rb-ui/frontend.kit-npm/src/frontend-kit/atoms/checkbox/checkbox.tsx
import * as React from 'react';
import classNames from 'classnames';
import { UiIcon } from '../uiIcon/uiIcon';

class CheckboxProps {
  id: string;
  labelCheckbox: string;
  isSelected?: boolean;
  isDisabled?: boolean;
  indeterminate?: boolean;
  state?: 'success' | 'warning' | 'error';
}

/**
 * @name        a-checkbox
 * @type        atom
 * @author      Experience One AG
 * @copyright   Robert Bosch GmbH
 *
 * @param       {string}    id        Unique ID for each checkbox
 * @param       {string}    labelCheckbox     Label to display
 * @param       {boolean}   isSelected        Select or unselect the input
 * @param       {boolean}   isDisabled        Wether or not the checkbox is disabled
 * @param       {boolean}   indeterminate     Wether or not the checkbox is indeterminate
 * @description
 * representation of checkboxes
 */

const Checkbox: React.FunctionComponent<CheckboxProps> = ({
  id,
  labelCheckbox,
  isSelected,
  isDisabled,
  indeterminate,
  state,
}: CheckboxProps) => {
  const idCheckbox = `checkbox-${id}`;

  const checkboxClass = classNames('a-checkbox', {
    'a-checkbox--indeterminate': indeterminate,
    [`-${state}`]: state,
  });

  return (
    <div className={checkboxClass}>
      <input
        type="checkbox"
        id={idCheckbox}
        defaultChecked={isSelected}
        disabled={isDisabled}
      />
      <label htmlFor={idCheckbox}>{labelCheckbox}</label>
      <UiIcon />
    </div>
  );
};

export { Checkbox, CheckboxProps };
