// @rb-ui/frontend.kit-npm/src/frontend-kit/atoms/textField/textField.tsx
import * as React from 'react';
import classNames from 'classnames';
import { UiIcon } from '../uiIcon/uiIcon';
import { Icon } from '../icon/icon';

class TextfieldProps {
  disabled?: boolean;
  id: string;
  placeholder?: string;
  label?: string;
  state?: 'success' | 'warning' | 'error';
  isPassword?: boolean;
}

/**
 * @name        a-text-field
 * @type        atom
 * @author      Experience One AG
 * @copyright   Robert Bosch GmbH
 *
 * @param       disabled      Wether or not the text field is disabled
 * @param       id              Unique ID for the text field
 * @param       placeholder     Placeholder of text field
 * @param       label           Label of text field
 * @param       state           Whether the state of text field is success, warning or error
 * @param       isPassword      Wether or not the text field is a password field

 * @description
 * representation of text field
 */

const TextField: React.FunctionComponent<TextfieldProps> = ({
  disabled,
  id,
  placeholder,
  label,
  state,
  isPassword,
}: TextfieldProps) => {
  const divClass = classNames(
    'a-text-field',
    { [`-${state}`]: state },
    { 'a-text-field--password': isPassword },
  );
  const inputType = isPassword ? 'password' : 'text';
  const isState = state || false;

  return (
    <div className={divClass}>
      {label && <label htmlFor={id}>{label}</label>}
      <input
        type={inputType}
        id={id}
        disabled={disabled}
        placeholder={placeholder}
      />
      {isPassword && (
        <div className="a-text-field__icon-password">
          <UiIcon titleText="LoremIpsum" name="watch-on" />
        </div>
      )}
      {isState && (
        <UiIcon className="a-text-field__icon-state" />
      )}
    </div>
  );
};

export { TextField, TextfieldProps };
