// @rb-ui/frontend.kit-npm/src/frontend-kit/atoms/notification/notification.tsx
import * as React from 'react';
import classNames from 'classnames';

import { Icon } from '../icon/icon';
import { Link } from '../link/link';
import { Button } from '../button/button';

import NotificationComponent from './index';

interface NotificationProps {
  variant?: 'banner' | 'text';
  type?: 'neutral' | 'success' | 'warning' | 'error';
  size?: 's';
  iconName?: string;
  notificationId?: string;
  children?: React.ReactNode;
  content?: React.ReactNode;
}

/**
 * @name        a-notification
 * @type        atom
 *
 * @param       variant         optional
 *                              if not passed 'bar' type will be generated.
 *                              Other variants are 'banner', 'text'
 * @param       type            optional
 *                              default will be neutral, other types are 'success', 'warning', 'error'
 * @param       size            optional
 *                              can be passed to get small notification, only available for bar type
 * @param       iconName        optional
 *                              icon name will only be used when type is neutral
 * @param       notificationId  optional, only needed when notification is banner variant
 * @param       children        optional, gets inserted into the inner div of the notification
 * @param       content         optional, replaces inner div of notification
 *
 * @description
 * representation of a notification
 */

const Notification: React.FunctionComponent<NotificationProps> = ({
  variant,
  type = 'neutral',
  size,
  iconName,
  notificationId,
  children,
  content,
}) => {
  const notificationLabelId = notificationId
    ? `${notificationId}-label`
    : 'notification-label-id';

  let selectedIconName: string = iconName;

  // construction of class names
  const notificationClass = classNames('a-notification', {
    [`a-notification--${variant}`]: variant,
    [`-${size}`]: size && !variant,
    [`-${type}`]: type,
  });

  /** selecting of icon, when type is not 'neutral'
   * prevent user to pass a different icon to notification
   * when type is different from 'neutral'
   */
  const iconMap = {
    success: 'alert-success',
    warning: 'alert-warning',
    error: 'alert-error',
  };

  if (type && type !== 'neutral') {
    selectedIconName = iconMap[type];
  }

  return (
    <div
      className={notificationClass}
      id={
        notificationId
          ? NotificationComponent.notificationId(notificationId)
          : null
      }
      role="alert"
      aria-labelledby={notificationLabelId}
    >
      {variant !== 'text' && selectedIconName && (
        <Icon iconName={selectedIconName} titleText="Loren Ipsum" />
      )}
      {content || (
        <div id={notificationLabelId} className="a-notification__content">
          {children || (
            <>
              At vero eos et accusam et justo duo dolores et ea rebum. Stet
              clita kasd gubergren, no sea{' '}
              <Link level="inline" label="Link sanctus" href="/" /> est Lorem
              ipsum dolor sit amet. At vero eos et accusam et justo duo dolores
              et ea rebum. Stet clita kasd gubergren, lorem ipsum dolor sit amet
            </>
          )}
        </div>
      )}
      {variant === 'banner' && (
        <Button action="close" type="close" label="close" />
      )}
    </div>
  );
};

export { Notification, NotificationProps };
